### ISSUE TITLE
Some problems to understand your definition of bad

### ISSUE THREAD

**Issue by varocarbas** [mitchellkrogza/nginx-ultimate-bad-bot-blocker#143]
You have included in [your bad-bot list](https://github.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/blob/master/_generator_lists/bad-user-agents.list) the ones (current/old versions) which are building my web domain ranking: RankingBot and RankingBot2.

They are clearly identified, pursue a neutral (even positive) goal, store minimal information, fully respect the robots.txt indications unless under extreme circumstances, etc. All this is clearly explained in [their own page](http://varocarbas.com/bot_ranking2/).

Can you please explain me how have you come to the conclusion that my clearly-identified, inoffensive and respectful bots are bad? Have you seen any kind of malfeasance about which I am not aware (bear in mind that I am their sole author and operator)? Or this is just some kind of because-I-say-it-so list?

**Comment by mitchellkrogza** [comment]
They are defined as follows. 

This system was built to protect my own web sites from:

- bad bots first
- secondly from referrer spam
- thirdly from a number of SEO companies selling services to people and aggressively scanning other people's web sites to help competitors gain insights on your SEO
- lastly from backlink scanners and ranking systems which provide similar insights 

This system was and is however designed in such a way that individual users who want to allow those User-Agents access can simply bypass them using the include files.

**Comment by varocarbas** [comment]
@mitchellkrogza 
Thanks for taking the time to finally answer me (8 days afterwards, not bad!).

First thing I want to highlight is that this is your list, your designations ("bad bots") and your decisions. Everything completely up to and to the people who relies on it. You are free to do whatever you wish. I will not insist or try to convince you about anything. This whole issue is about clearly describing my position regarding what I think about my bots (or similar) being included in any list of this kind. 

I understand that you consider my bots part of your fourth point ( `lastly from backlink scanners and ranking systems which provide similar insights`) and, as such, I see various issues:

- Most of web bots focus on text/links. For example, all the ones of search engine, directories, indexers, etc. I didn't see any mention to bots from Google or Bing or Yandex, etc. which are extremely much more likely of having a higher bothering impact (bigger companies with much more resources analysing many sites over and over).
- The other problem of censoring these bots is that your site stops being recognised by whatever rankings/search engines, what is usually against the intend of most of users. It does make sense to avoid certain pages to be analysed but rarely the whole site, as far as one of the points of being internet is being searchable. For example, all the sites restricting the access to my bots are provoking them to ignore their information, marginally reducing the overall quality of the ranking and seriously affect the impact on it of the given site.
- You seem to assume that my bots (or I) have anything to do with ranking/SEO when this is extremely far away from the reality. I built this whole ranking (the bots, the underlying system, backups, communications, etc.) to showcase my programming skills. More specifically, my efficiency-focused side as far as these whole system is working under very limited hardware resources (e.g., all the bots are running from a pretty crappy old computer); also my custom-algorithm building (developed them from scratch) and scalability (running for over 1 year already finding thousands of new domains on a daily basis) skills. If you go to any page (out of the virtually infinite number of them) of the associated ranking, you would see that there is no links. The whole point is showing positions by applying as generically objective rules as possible. I don't earn anything from this, other than over work, giving a slightly bad impression if the results don't look good enough and, eventually, having to deal with generic prohibitions like what your list represents.

In any case, what bothers me the most is the "bad" label which seem to issue quite arbitrarily and easily. Your first category are "bad bots", which I understand that are malware or similar. Your second category is spam that also sounds quite bad. But from that point downwards, things start getting a bit weird. I am not precisely a defender of marketing-whatever, but nothing of this seems to belong to the same category than the first two ones. At least not when being as less invasive as possible, the case of my bots which only store minimal, virtually anonymous, information. Nothing of this seems fair, logical or even sensible. Much less by bearing in mind the small impact of my bots (over 100 times visited important sites and 1-10 smaller ones during a whole year?), what makes me wonder about the exact technique used to populate that list.

In summary, my problem isn't sites allowing/restricting my bots (it would mostly be bad for them), but they being unfairly included in a list on these lines. Hopefully, I my point is clear enough.

**Comment by mitchellkrogza** [comment]
@varocarbas no problem and point taken, thanks for your valuable input and I will certainly review these again. I was away for a week but normally I respond within a day or two though.

**Comment by varocarbas** [comment]
@mitchellkrogza  OK, thanks. No problem.

**Comment by corbolais** [comment]
FTR, I chose UBBB for what I want to achieve: exclusion of bots. Therefore, I explicitely welcome the inclusion of ***any*** bot in the list. I do ***not*** want bots on my sites. I think bots are an evil per se and render the net all the worse. It would be better w/o any bot whatsoever.

I think this position is clear.

@mitchellkrogza keep up the good work. 

@varocarbas stop developing software, and your programming skills do not matter when it comes to my sites. Bad reasoning on top of it.

**Comment by varocarbas** [comment]
@corbolais  `FTR, I chose UBBB for what I want to achieve: exclusion of bots. `

Extremely naive statement. None of the most active bots are in that list, that is, the ones from sites like Google.

`I think bots are an evil per se and render the net all the worse`

So, you aren't using search engines I understand, because without bots there will be no search engine. Or any even slightly complex understanding of anything. Bots are pieces of software performing automated actions, they are one of the basis of internet. The fact that some people build pieces of software performing dishonest actions and call them 'bots' doesn't mean that they do the same than any other piece of software with that name. Are you aware about the importance of context/properly understanding anything?

`stop developing software, and your programming skills do not matter when it comes to my sites. Bad reasoning on top of it.`

Why? Have I asked you to stop having a website? My programming skills are promoted by my bots visiting your site or not. In fact, my bots will not visit your site if you don't want (no need of this list, just include its name in your robots.txt or tell me your sites right away). BTW, my bots collect in each visit much lower amount of information than any person (or most of other bots): pretty much counters of links from sites to sites, nothing else and nothing more. 

You are very aggressively defending a position about which you don't seem to have too much knowledge.

**Comment by ScrewLooseDan** [comment]
> I built this whole ranking to showcase my programming skills

This is why I'll continue to block your bot regardless if the author decides to exclude you from his list.

Best of luck.

**Comment by varocarbas** [comment]
@ScrewLooseDan  Completely up to you, but you don't seem to have got my point. Building a comprehensive software system (formed by the bots, ranking algorithms, communication subsystems, etc. all this under very restricted hardware conditions) is the promotion, working fine I mean. 

Also I am not sure why this is worse than having built it for a direct profit? Someone does something which you might eventually enjoy and his motivation is doing it properly (to showcase his skills), but you would have preferred he getting a direct benefit? Kind of weird, but good luck to you to (I don't need it, BTW).

**Comment by thezoggy** [comment]
I personally view unwanted requests to my site as a resource drain. I have a limited bandwidth and low powered vps.. so I'd rather cater towards requests that are legit.

**Comment by varocarbas** [comment]
@thezoggy  As said to others before, you are free to do whatever you wish. This isn't about convincing anyone to allow my bots, but to complain about they having been unfairly tagged as "bad".

In any case, do you mind to elaborate on your "being legit"? I have explained in detail the purpose of my bots and their low impact everywhere (actually, I found you including a so modest bots in your list kind of incredible! How could you know about something which has a so small activity? The chances are very low). Their main goal is being as objective and absolute-correctness-focused as possible. No nepotism. No paid improvements. Not even dishonest inter-linking. Whatever domain ranks higher is because it deserves it (and/or I made an error). Their impact on any site is negligible, much less lately because I am speeding them up (the less time they spend in a site, the lower load in my server and, theoretically, the more comprehensive are the conclusions; in this way, they find a lower number of new domains, but after having already over 16M this isn't a concern anymore). If one of them visits your site, it would go to a around 20-100 different pages very quickly (looking for links and out), will not store any information about you and will move to the next one. Any user spending 5 minutes in one of your pages would consume way much more resources than my bots. Any bot of a major site systematically visiting everything and collecting everything would have a way much higher impact that my bots. I don't see any of these bots in your list. Why aren't apply the same rules to everyone (or even more drastically to the more aggressive alternatives) and ban the bots from Google and company. Or, at least, the ones from secondary indexing/search engines. I get lots of visits from Czech, Chinese, Russian search engines about which I don't care at all. All of them with a much higher impact on my site that what my bots will have during a whole year (visiting me 1-3 times).

So, do you mind to elaborate a bit on your position or shall I feel offended with you implying that my bots aren't "legit", when ironically you are the ones who have unilaterally started this whole discussion because of unfairly calling me and work "bad" and not saying me a word (knew about it by pure accident). I haven't said a word about what I think about the "legit" status of your list, should I start?

**Comment by mitchellkrogza** [comment]
@varocarbas you are taking too much offence to the general and very commonly used term "bad bots". The word bad is not intended to defame you or anyone it is merely a term as old as the internet itself. 

**Undesirable** may be a better word but almost every User-Agent blocking system has adopted the word bad. Perhaps even the words **"trusted"** or **"untrusted"** might be better terms. 

I am going to allow your bot to crawl some of my sites for the next few days to help make a decision.

**Comment by varocarbas** [comment]
@mitchellkrogza 
It is not exactly offended, but feeling that my work/effort is being unfairly mistagged. All the words you are using sound equally non-applicable to my bots IMHO. A word like "irrelevant" seems much more correct. Bear in mind that my whole activity is online/remote based and I put a major focus on issues like solid principles, objective correctness, etc. You can call me poor if you wish, or even a bit anti-social, but censoring the quality or honesty of my work is kind of a problem.

Anyway, thanks for giving them a shot. But you might need more than just a few days for any of my bots to reach any of your sites. Just to have a rough idea about the amount of information involved and the limited resources: at the moment, it takes them around 3-4 months to go from the firstly-found domain to the last one (almost 17M now) even though currently being in high speed mode. And this is just a very small part of all the available domains, you might even have some websites which they haven't found yet. This is an issue which puzzles me about all this: how could a so modest activity have triggered any kind of alarm?!

I look forward to your impressions and final decision.

**Comment by mitchellkrogza** [comment]
A search through 13 months of my server logs spanning 36 web sites reveals only 109 hits. 

I will remove this from the main list of bots in next commit and continue to monitor the activities of this crawler.

```grep -o 'RankingBot' /var/log/nginx/*-access.log* | wc -l 
109``

```grep -o 'RankingBot2' /var/log/nginx/*-access.log* | wc -l 
109``

Users who wish to continue blocking RankingBot may continue to do so by adding

`"~*\bRankingBot\b"			3;`

into the `blacklist-user-agents.conf` include file.

**Comment by varocarbas** [comment]
@mitchellkrogza  Thanks.

**Comment by corbolais** [comment]
@mitchellkrogza Hey there, so all it just takes is a nag that keeps coming back until you comply?
Look at the sheer word count this guy spews out. That alone would warrant a deny imo. ubbb for me was about blocking unwanted bots including nagger's bots.
You ask about opinion, you got a couple o' "nay!" and then you go "okay! happy to comply!". Seriously?
Pathetic.

**Comment by mitchellkrogza** [comment]
@corbolais I look at various aspects before considering removal. I am monitoring it and will be happy to re-include it again if needs be. If you feel it should remain here please motivate, for now I have not seen enough counts of it in my logs but that does not mean others like you may be seeing more inflated hit counts so, please can you contribute any findings.

**Comment by StarkRavingZA** [comment]
IMHO it should remain on the list, it falls into the same category as many other SEO companies and user agents listed here already.

**Comment by varocarbas** [comment]
@mitchellkrogza  Sorry if I have misunderstood something, but your list still includes references to my bots. 

A short while after my last comment, I did confirm that they were deleted from your list but now they are back?! Is this normal within your bot-removal process? Or shall I understand that you changed your mind? Apparently for no reason or is there anything you want to share with me?

**Comment by mitchellkrogza** [comment]
@varocarbas users of the blocker have requested it's re-addition. The only way for users to allow the bot through is to add it in their own custom whitelists that the blocker provides.

**Comment by mitchellkrogza** [comment]
https://github.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/commit/9ec48efe9e183ae5fe50547d9c71a1f45725bf76
